dataclasses_json/core.py

```diff
@@ -314,14 +314,16 @@ def _decode_generic(type_, value, infer_missing):
             type_options = _get_type_args(type_)
             res = value  # assume already decoded
             if type(value) is dict and dict not in type_options:
-                # FIXME if all types in the union are dataclasses this
-                #  will just pick the first option -
-                #  maybe find the best fitting class in that case instead?
+                changed = False
                 for type_option in type_options:
                     if is_dataclass(type_option):
-                        res = _decode_dataclass(type_option, value, infer_missing)
-                        break
-                if res == value:
+                        try:
+                            res = _decode_dataclass(type_option, value, infer_missing)
+                            changed = True
```

Comment:
The changed part is not necessary here? As I understand you just add try...except?

Reply:
We only mark `changed` as true when a decode is successful (when no exception is thrown), so that if no decode attempts are successful a warning will be emitted.